core/src/main/scala/com/pingcap/tispark/auth/TiAuthorization.scala

```diff
@@ -0,0 +1,282 @@
+import java.util.concurrent.{Executors, ScheduledExecutorService, TimeUnit}
+import scala.collection.JavaConverters
+import scala.util.control.Breaks.{break, breakable}
+
+case class TiAuthorization private (parameters: Map[String, String], tiConf: TiConfiguration) {
+
+  private var tiDBJDBCClient: TiDBJDBCClient = _
+
+  private val scheduler: ScheduledExecutorService =
+    Executors.newScheduledThreadPool(1)
+
+  val globalPriv: AtomicReference[List[MySQLPriv.Value]] =
+    new AtomicReference(List())
+
+  val databasePrivs: AtomicReference[Map[String, List[MySQLPriv.Value]]] =
+    new AtomicReference(Map())
+
+  val tablePrivs: AtomicReference[Map[String, Map[String, List[MySQLPriv.Value]]]] =
+    new AtomicReference(Map())
+
+  val task: Runnable = () => {
+    val privs = getPrivileges
+    globalPriv.getAndSet(privs.globalPriv)
+    databasePrivs.getAndSet(privs.databasePrivs)
+    tablePrivs.getAndSet(privs.tablePrivs)
+  }
+
+  {
+    TiAuthorization.dbPrefix = tiConf.getDBPrefix
+    val option = new TiDBOptions(parameters)
+    try {
+      this.tiDBJDBCClient = new TiDBJDBCClient(TiDBUtils.createConnectionFactory(option.url)())
+    } catch {
+      case e: Throwable => {
+        logger.error(f"Failed to create tidb jdbc client with url ${option.url}", e)
+        System.exit(-1)
+      }
+    }
+
+    task.run()
+    scheduler.scheduleWithFixedDelay(task, refreshInterval, refreshInterval, TimeUnit.SECONDS)
+  }
+
+  def getPrivileges: PrivilegeObject = {
+    var input = JavaConverters.asScalaBuffer(tiDBJDBCClient.showGrants).toList
+
+    /** TODO: role-based privilege
+     *
+      * Show grants using more than two roles return incorrect result
+     * https://github.com/pingcap/tidb/issues/30855
+     *
+      * val roles = extractRoles(input)
+     * input =
+     * if (roles.nonEmpty)
+     * JavaConverters
+     * .asScalaBuffer(
+     * tiDBJDBCClient.showGrantsUsingRole(roles.asJava)
+     * ).toList
+     * else input
+     */
+
+    parsePrivilegeFromRow(input)
+  }
+
+  def getPDAddress(): String = {
+    try {
+      tiDBJDBCClient.getPDAddress
+    } catch {
+      case e: Throwable =>
+        throw new IllegalArgumentException(
+          "Failed to get pdAddress from TiDB, please make sure user has `PROCESS` privilege on `INFORMATION_SCHEMA`.`CLUSTER_INFO`",
+          e)
+    }
+  }
+
+  def checkGlobalPiv(mySQLPriv: MySQLPriv.Value): Boolean = {
+    val privs = globalPriv.get()
+    if (privs.contains(MySQLPriv.AllPriv) || privs.contains(mySQLPriv)) true
+    else false
+  }
+
+  def checkDatabasePiv(db: String, mySQLPriv: MySQLPriv.Value): Boolean = {
+    val privs = databasePrivs.get().getOrElse(db, List())
+    if (privs.contains(MySQLPriv.AllPriv) || privs.contains(mySQLPriv)) true
+    else false
+  }
+
+  def checkTablePiv(db: String, table: String, mySQLPriv: MySQLPriv.Value): Boolean = {
+    val privs =
+      tablePrivs.get().getOrElse(db.trim, Map()).getOrElse(table.trim, List())
```

Comment:
could you please add a comment for this? what is the behavior of privs when tablePrivs/databasePrivs exists/not exists on the current table & database

Reply:
done